draft-ietf-tls-tls13.md

```diff
@@ -4375,6 +4375,109 @@ currently deployed and properly configured TLS implementations.
 %%### Ticket Establishment
 %%### Updating Keys
 
+# State Machine
+
+This section provides a summary of the legal state transitions for the
+client and server handshakes. State names have no formal meaning but
+are provided in all capitals (e.g., START) for ease of comprehension.
+Messages which are sent only sometimes are indicated in [].
+
+## Client
+
+~~~~
+                +-----> START
+           Recv |        |
+            HRR |        | Send CH
+                |        v
+       +->      +---- WAIT_SH            
+       |                 |                 
+       |                 | Recv SH         
+       |                 |                 
+       |                 V                 
+       |              WAIT_EE            
+       |                 |                 
+       |                 | Recv EE         
+       |                 |                 
+       |          +------+------+          
+       |          |             |              
+       |   SH+PSK |             | SH
+       |          |             v
+       |          |         WAIT_CERT_CR
+   Can |          |          |       |
+  Send |          |          |       |
+ Early |          |     Recv |       | Recv CR
+  Data |          |     Cert |       |
+  Here |          |          |       v
+       |          |          |    WAIT_CERT
+       |          |          |       |
+       |          |          |       | Recv Cert
+       |          |          v       v
+       |          |           WAIT_CV
+       |          |              |
+       |          |              | Recv CV
+       |          |              v
+       |          +-------> WAIT_FINISHED
+       |                         |
+       |                         | Recv Finished
+       |                         |
+       |                         v
+       +->                  [Send EOED]
+                            [Send Cert + CV]
+                            Send Finished
+                                 |
+                                 v
+                             CONNECTED
+~~~~
+
+## Server
+
+~~~~
+         +-----> START
+ Recv CH |      |  |
+ Send HRR|      |  | Recv CH
```

Comment:
This part seems kind of confusing. If you rename NEGOATIATED to, idk, RECV_CH, then you can make a "Send HRR" transition from there to START without duplicating "Recv CH". You could also keep both RECV_CH and NEGOTIATED and have a "Negotiate" or "Pick parameters" or whatever transition between the two.

```
                 START <--------+
                   |            |
                   | Recv CH    | Send HRR
                   |            |
                   v            |
                RECVED_CH ------+
                   |
                   | Select params
                   |
                   v
                NEGOTIATED
                   |
                   | Send SH
                  ...
```

IMO it makes it clearer (YMMV), with the added bonus that it makes the machine somewhat more deterministic. Con: it adds an additional state.